src/rules/no_unexpected_multiline.rs

```diff
@@ -0,0 +1,322 @@
+      return;
+    }
+
+    let tag = &tagged_tpl.tag;
+    let tag_end_loc =
+      self.context.source_map.lookup_char_pos((&*tag).span().hi());
+
+    let quasi = &tagged_tpl.quasis[0];
+    let quasi_start_loc =
+      self.context.source_map.lookup_char_pos(quasi.span().lo());
+    if tag_end_loc.line != quasi_start_loc.line {
+      self.context.add_diagnostic(
+        quasi.span(),
+        "no-unexpected-multiline",
+        "Unexpected newline between template tag and template literal",
+      );
+    }
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use crate::test_util::*;
+
+  #[test]
+  fn no_unexpected_multiline_valid() {
+    assert_lint_ok::<NoUnexpectedMultiline>(
+      r#"var foo = bar;
+(1 || 2).baz();
+
+var foo = bar
+;(1 || 2).baz()
+
+foo<string>
+("").length;
+
+(foo
+)(bar);
+
+// (foo).callback
+// ?.
+// (bar)
+
+var hello = 'world';
+[1, 2, 3].forEach(addNumber);
+
+(array
+)[1]
+
+var hello = 'world'
+void [1, 2, 3].forEach(addNumber);
+
+function foo() { return ""; }
+foo
+().length
+
+function foo<T>(bar: T): T { return bar; }
+`hello`
+
+let x = function() {};
+`hello`
+
+let tag = function() {}
+tag `hello`
+
+let a = b/
+abc/g;
+
+let x = a
+/foo/ g"#,
+    );
+  }
+
+  #[test]
+  fn no_unexpected_multiline_invalid() {
+    assert_lint_err_on_line::<NoUnexpectedMultiline>(
+      r#"var foo = bar
+(1 || 2).baz();"#,
+      2,
+      1,
+    );
+    assert_lint_err_on_line::<NoUnexpectedMultiline>(
+      r#"foo
+<string>
+("")"#,
+      2,
+      1,
+    );
+    assert_lint_err_on_line::<NoUnexpectedMultiline>(
+      r#"(foo)
+(abc).length"#,
+      2,
+      1,
+    );
+    assert_lint_err_on_line::<NoUnexpectedMultiline>(
+      r#"foo(bar
+(x), baz)"#,
+      2,
+      1,
+    );
+    /*assert_lint_err_on_line::<NoUnexpectedMultiline>(
+          r#"(foo).callback
+    (bar)?.baz"#,
+          2,
+          1,
+        );*/
+    assert_lint_err_on_line::<NoUnexpectedMultiline>(
+      r#"var hello = 'world'
+[1, 2, 3].forEach(addNumber);"#,
+      2,
+      1,
+    );
+    assert_lint_err_on_line::<NoUnexpectedMultiline>(
+      r#"var a = b
+/
+abc/g-a"#,
+      2,
+      1,
+    );
+    assert_lint_err_on_line::<NoUnexpectedMultiline>(
+      r#"let x = foo
+/regex/g.foo(bar)"#,
+      2,
+      1,
+    );
+  }
+
+  #[test]
+  fn no_unexpected_multiline_invalid_tagged_tpl() {
+    assert_lint_err_on_line::<NoUnexpectedMultiline>(
+      r#"let x = function() {}
+`hello`"#,
+      2,
+      1,
+    );
+    assert_lint_err_on_line::<NoUnexpectedMultiline>(
+      r#"let x = function() {}
+x
+`hello`"#,
+      3,
+      1,
+    );
+    assert_lint_err_on_line::<NoUnexpectedMultiline>(
+      r#"(foo)
+`hello`"#,
+      2,
+      1,
+    );
+    assert_lint_err_on_line::<NoUnexpectedMultiline>(
+      r#"bar<string>
+`${x} hello`"#,
+      2,
+      1,
+    );
+    assert_lint_err_on_line::<NoUnexpectedMultiline>(
+      r#"const x = aaaa<
+  test
+>/*
+test
+*/`foo`"#,
+      5,
+      3,
+    );
```

Comment:
Nice tests, are they ported from ESLint?

Reply:
In part, yes. That last one for example is from ESLint. Probably would not have come up with this by myself.